### PR TITLE
Switch to go modules

### PR DESCRIPTION
**Pull request by ialidzhikov**
**What this PR does / why we need it**:

The PR adopts go modules to manage the vendor folder in gardener/gardener and to define go.mod module files.

More information about the motivation and goals for go modules adoption can be found at the corresponding kubernetes/enhancements - https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/2019-03-19-go-modules.md

Conceptually, the steps I executed:
```bash
# 1. init go modules
$ go mod init github.com/gardener/gardener

# 2. use replace directive in go.mod to make sure that
# the concrete desired version will be picked up

# 3. introduce tools.go (required for ./hack/generate-code)

# 4. adapt Makefile, .ci/, hack/ to be module aware and to use -mod=vendor

# 5. revendor
$ make revendor

# 6. remove dep leftovers
$ rm Gopkg.toml Gopkg.lock

# 7. update docs/
```

Ref #1096.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```noteworthy developer
`dep` is replaced by `go mod`. `github.com/gardener/gardener` now publishes go module files containing dependency version information. 
```

### CODE REVIEW

**Comment by zanetworker** [comment]
@ialidzhikov this is still WIP right?

**Comment by ialidzhikov** [comment]
> @ialidzhikov this is still WIP right?

Yep. For now I will put a `/hold`, once it is ready for review I will put `/hold cancel`.

/hold

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,132 @@
+module github.com/gardener/gardener
+
+go 1.12
+
+require (
```
Comment (rfranzke) [inline go.mod:5]: Is it possible to not include the indirect dependencies here? It's very hard to maintain this way.

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	k8s.io/kube-aggregator v0.0.0-20190314000639-da8327669ac5
+	k8s.io/kube-openapi v0.0.0-20190320154901-5e45bb682580
+	k8s.io/utils v0.0.0-20190506122338-8fab8cb257d5
+	sigs.k8s.io/controller-runtime v0.2.0-beta.2
+	sigs.k8s.io/structured-merge-diff v0.0.0-20190302045857-e85c7b244fd2 // indirect
+)
+
+replace (
+	github.com/coreos/etcd => github.com/coreos/etcd v3.3.12+incompatible
+	github.com/gardener/external-dns-management => github.com/gardener/external-dns-management v0.0.0-20190220100540-b4bbb5832a03 // 0.1.0
```
Comment (rfranzke) [inline go.mod:115]: Why is this replaced/duplicated here? The same version does also exist in the above list?
Can we get rid of this whole section here?

File: go.mod (hunk 1 above)
Comment (ialidzhikov) [reply go.mod:5]: Currently I am not aware of such way.
Ref https://github.com/golang/go/wiki/Modules#why-does-go-mod-tidy-record-indirect-and-test-dependencies-in-my-gomod

File: go.mod (hunk 2 above)
Comment (ialidzhikov) [reply go.mod:115]: I used replace directives to make sure that the concrete desired version will be picked up. You can note that for some of the them the version in `require` differs from the version in `replace`.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**